### PR TITLE
Fix #6542: Pickle line sizes in TASTy

### LINKED ISSUE #6542: Line numbers from TASTy
When we have a position on a tree the was loaded from TASTy we have no way to compute line numbers and columns as we only have the offset but no source file.

We have identified the following scenarios where we require the line numbers from TASTy files:
* Error reporting on code inlined from a library for which we do not have the sources (see #6538 and #6412)
* Recompiling from TASTy without sources available. The line numbers are required for the bytecode stack trace information.

So far the most robust solutions that we have are:
* Keep the line sizes in the TASTy file. Has a small memory footprint.
* Keep the actual source in the TASTy file. This would have a large footprint but would also give the possibility of printing better errors for inlined positions.

Which solution should be implemented?

### PR DESCRIPTION
**Pull request by nicolasstucki**
**This description is outdated, see [comment for second commit bellow](https://github.com/lampepfl/dotty/pull/10363#issuecomment-731563022).**

#### Line Sizes section

This PR add a new section called `LineSizes` to the tasty format. ~This section starts with an `Int` containing the number of lines followed by an~ This section h sequence of `Int` with the size of each line. It requires around one byte for each line (2 bytes if line longer than 127 characters) and extra section info.

This information is then used to compute the line offsets in `SourceFile` to be able to translate `Span` offsets into line/column numbers.

The jar of the standard library increased form `6,647,453 bytes` to `6,893,064 bytes`. This is roughly a `3.7%` increase in file size. 

For sources in `scala.collection.immutable` the _non-compressed_ TASTy files increased cumulative sizes increased from `964,421 bytes` to `1,100,330 bytes` (filesystem wise in Mac). This is roughly a `14%` increase in file size. 

Here is a subset of byte sizes of some sections (not including headers and names)

|  Source | AST   | Positions  |  Line sizes |  Docs | 
|---|---|---|---|---|
|  List.tasty |  7343 | 7671  |  735 | 4609  |
| Vector.tsaty  | 5542  | 5192  | 2617  |  2158  |  
|  Map.tasty | 14939  |  13566 |  844 | 1143  |
| SortedSet.tasty  | 509  | 389  | 64  |  263 |
|  Set.tasty | 6657  | 6237  | 456  | 675  |
|  Range.tasty | 9321  | 8830  | 798  | 4531  |
|  Iterable.tasty | 413  | 328  | 45  | 205  |
| ArraySeq.tasty | 15879 | 14674 | 851 | 1917 |

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/tasty/LineSizesPickler.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+  import ast.tpd._
+  val buf: TastyBuffer = new TastyBuffer(5000)
+  pickler.newSection("LineSizes", buf)
+
+  def pickleLineNumbers(source: SourceFile): Unit = {
+    val content = source.content()
+
+    val numLines = content.count(_ == '\n')
+    buf.writeInt(numLines + 1)
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/core/tasty/LineSizesPickler.scala:29]: I might be able to not pickle this one if we look at the size of the section when unpickling

**Comment by nicolasstucki** [comment]
test performance please

**Comment by nicolasstucki** [comment]
test performance please

**Comment by smarter** [comment]
I don't know if it would be better or worse in term of size, but it might be worth trying to store the line number as part of the PositionPickler, using delta-coding like it's already doing for the start/end/span.

**Comment by nicolasstucki** [comment]
The delta coding wouldn't save us anything as we are already close to one byte per line. We could save on the section name and size of we join them.

**Comment by nicolasstucki** [comment]
The second commit puts the line sizes directly in the `Positions` section. In this version, we pickle the number of lines followed by the line sizes at the start of the `Positions` section.

This simplifies considerably the code required to implement this addition.

The jar of the standard library increased from `6,647,453 bytes` `6,851,555 bytes` (instead of the previous `6,893,064 bytes`). This is a `3.1%` (instead of `3.7%`) increase in jar file size.

**Comment by nicolasstucki** [comment]
test performance please

**Comment by smarter** [comment]
> The delta coding wouldn't save us anything as we are already close to one byte per line.

Since the lien number delta between two trees should be 0 or 1 in most cases, it's possible we could find a better encoding that uses less than one byte per line, but I'm not sure.

**Comment by sjrd** [comment]
One thing to pay attention to, if not already done: the different kinds of newlines and how they interact with offsets. In particular, the Windows `\r\n` takes 2 bytes, so it's a single line jump but with a shift of 2 bytes/Chars in offsets.

**Comment by liufengyun** [comment]
It would be good to add neg test for inlined trees from separately compiled files to make sure it points to the correct source file and line.

**Comment by nicolasstucki** [comment]
Added a regression test in `project/scripts/cmdTests`.

File: compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala [hunk 2]
```diff
@@ -1366,9 +1364,9 @@ class TreeUnpickler(reader: TastyReader,
       val path = sourcePathAt(addr)
       if (path.nonEmpty) {
         val sourceFile = ctx.getSource(path)
-        lineSizesUnpicklerOpt match
-          case Some(lineSizesUnpickler) =>
-            sourceFile.setLineIndices(lineSizesUnpickler.lineIndices)
+        posUnpicklerOpt match
+          case Some(posUnpickler) =>
+            sourceFile.setLineIndicesFromLineSizes(posUnpickler.lineSizes)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1369]: There is a potential issue here: the relative path of a file is not a unique identifier of the file, there might be path conflicts in the Scala ecosystem.

If we also store the hash of source files, that would be better.

**Comment by nicolasstucki** [comment]
Could someone with a Windows machine help me check if the following behaves the same?

<details>
<summary>Reproduction steps</summary>

```
# cp tests/neg/i6371/A_1.scala A.scala     
# cp tests/neg/i6371/B_2.scala B.scala    
# mkdir outA
# sbt   
sbt:scala3> scalac -d outA A.scala
sbt:scala3> scalac -print-tasty outA/A.tasty
Names:
  ...
  ...

Trees:
start = Addr(0), base = 300, current = Addr(0), end = Addr(146)
146 bytes of AST, base = Addr(0)
  ...
  ...

 97 position bytes:
   lines: 7
   line sizes: 10, 42, 18, 13, 3, 1, 0
   positions:
         0: 0 .. 92
         5: 0 .. 0
         8: 0 .. 0
        24: 0 .. 92
        27: 13 .. 90
        34: 7 .. 7
        41: 13 .. 13
        45: 13 .. 13
        49: 13 .. 13
        54: 7 .. 7
        58: 7 .. 7
        67: 7 .. 7
        71: 7 .. 7
        76: 13 .. 90
        79: 28 .. 34
        82: 31 .. 34
        88: 37 .. 41
        96: 44 .. 90
        98: 44 .. 45
       100: 58 .. 72
       104: 63 .. 64
       110: 66 .. 69
       116: 70 .. 72
       117: 77 .. 86
       119: 82 .. 83
       125: 84 .. 86


 0 comment bytes:

sbt:scala3> scalac -classpath outA -d outA B.scala
-- Error: B.scala:2:7 ----------------------------------------------------------
2 |  A.foo("aa") // error
  |  ^^^^^^^^^^^
  |this case is unreachable since type String is not a subclass of class Integer
  | This location contains code that was inlined from A.scala:3
1 error found

sbt:scala3> exit
# rm A.scala 
# sbt
sbt:scala3> scalac -classpath outA -d outA B.scala
-- Error: B.scala:2:7 ----------------------------------------------------------
2 |  A.foo("aa") // error
  |  ^^^^^^^^^^^
  |this case is unreachable since type String is not a subclass of class Integer
  | This location contains code that was inlined from A.scala:3
1 error found
```
</details>

**Comment by liufengyun** [comment]
> Could someone with a Windows machine help me check if the following behaves the same?
> 
> Reproduction steps

I just sent you by email.

**Comment by nicolasstucki** [comment]
Executing the test on Windows yielded the same result
<details>
  <summary>Reproduction steps (Windows)</summary>

```
dotty-2>copy tests\neg\i6371\A_1.scala A.scala
        1 file(s) copied.
dotty-2>copy tests\neg\i6371\B_2.scala B.scala
        1 file(s) copied.
dotty-2>mkdir outA
dotty-2>sbt
sbt:scala3> scalac -d outA A.scala

Names:
   
   ...

Trees:
start = Addr(0), base = 300, current = Addr(0), end = Addr(146)
146 bytes of AST, base = Addr(0)

   ...

 97 position bytes:
   lines: 7
   line sizes: 10, 42, 18, 13, 3, 1, 0
   positions:
         0: 0 .. 92
         5: 0 .. 0
         8: 0 .. 0
        24: 0 .. 92
        27: 13 .. 90
        34: 7 .. 7
        41: 13 .. 13
        45: 13 .. 13
        49: 13 .. 13
        54: 7 .. 7
        58: 7 .. 7
        67: 7 .. 7
        71: 7 .. 7
        76: 13 .. 90
        79: 28 .. 34
        82: 31 .. 34
        88: 37 .. 41
        96: 44 .. 90
        98: 44 .. 45
       100: 58 .. 72
       104: 63 .. 64
       110: 66 .. 69
       116: 70 .. 72
       117: 77 .. 86
       119: 82 .. 83
       125: 84 .. 86


 0 comment bytes:

sbt:scala3> scalac -classpath outA -d outA B.scala
-- Error: B.scala:2:7 ----------------------------------------------------------
2 |  A.foo("aa") // error
  |  ^^^^^^^^^^^
  |this case is unreachable since type String is not a subclass of class Integer
  | This location contains code that was inlined from A.scala:3
1 error found

sbt:scala3> exit

dotty-2>del A.scala

dotty-2>sbt
-- Error: B.scala:2:7 ----------------------------------------------------------
2 |  A.foo("aa") // error
  |  ^^^^^^^^^^^
  |this case is unreachable since type String is not a subclass of class Integer
  | This location contains code that was inlined from A.scala:3
1 error found
```
</detaills>

File: compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1369]: How would you use that hash?

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1369]: Maybe `ctx.getSource(path)` can take hash as an argument? I haven't thought through the details.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1369]: The hash would be on the contents of the source? Then we would need to read all the sources eagerly when we add them to the context. That may be quite expensive.

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1369]: When we compile files, we already have the file contents in memory.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1369]: Actually, it is only loaded when we first access something the depends on the source https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/util/SourceFile.scala#L43.

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1369]: When we compile Scala files, the contents for the Scala files will be forced.

File: compiler/src/dotty/tools/dotc/core/tasty/PositionPickler.scala [hunk 3]
```diff
@@ -32,7 +32,21 @@ class PositionPickler(
     (addrDelta << 3) | (toInt(hasStartDelta) << 2) | (toInt(hasEndDelta) << 1) | toInt(hasPoint)
   }
 
-  def picklePositions(roots: List[Tree], warnings: mutable.ListBuffer[String]): Unit = {
+  def picklePositions(source: SourceFile, roots: List[Tree], warnings: mutable.ListBuffer[String]): Unit = {
+    /** Pickle the number of lines followed by the length of each line */
+    def pickleLineOffsetts(): Unit = {
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/tasty/PositionPickler.scala:37]: Spelling: LineOffsets

**Review by odersky — APPROVED** [review]
Otherwise LGTM

**Comment by nicolasstucki** [comment]
@odersky I also changed `writeInt`/`readInt` to `writeNat`/`readNat`, not only in the docs.

**Pull request merged**